lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp

```diff
@@ -650,6 +650,45 @@ BlockedEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape) const {
   return combineCtaCgaWithShape(ctaLayout, getCTALayout(), shape);
 }
 
+std::optional<LinearLayout>
+fmaDotToLinearLayout(DotOperandEncodingAttr operandLayout,
+                     ArrayRef<int64_t> shape) {
+  int rank = shape.size();
+  auto blocked = cast<BlockedEncodingAttr>(operandLayout.getParent());
+  MLIRContext *ctx = operandLayout.getContext();
+
+  // TODO: introduce registerOrder or use getOrder(operandLayout)
+  // Currently this order is used in legacy converter, because we do not
+  // have access to full dot operand layout, only parent part.
+  auto regOrder = blocked.getOrder();
+  // TODO: use operandLayout.getThreadOrder()
+  auto threadOrder = blocked.getThreadOrder();
+  auto warpOrder = blocked.getWarpOrder();
```

Comment:
I feel like there's something wrong here. Have you tested a warp shape of `[2, 2]` with more than 1 warp on the k dimension?

Reply:
My point is that warps have to be broadcasted along the k dimension

Reply:
Agreed. I think you have to use `warpsDotOperand` here.

Reply:
> Have you tested a warp shape of [2, 2] with more than 1 warp on the k dimension?

Yes, this converter works with any warp shape. The trick is I create "thread" part of layout using whole K dimension, so any number of warps or threads across k dimension will be squeezed in `combineCtaCgaWithShape` call

Reply:
> Agreed. I think you have to use warpsDotOperand here.

I thought about this, the only reason I choose to go without it for now is aestetic:

At this moment cta tile constructions looks like this:
```
  LinearLayout ctaLayout = identityStandardND(kReg, threadSize, regOrder)
                               .transposeOuts(repDimNames) *
                           identityStandardND(kLane, threadShape, threadOrder)
                               .transposeOuts(repDimNames) *
                           identityStandardND(kWarp, warpShape, warpOrder)
                               .transposeOuts(repDimNames);
```

with `warpsDotOperand`:
```
  LinearLayout ctaLayout = identityStandardND(kReg, threadSize, regOrder)
                               .transposeOuts(repDimNames) *
                           identityStandardND(kLane, threadShape, threadOrder)
                               .transposeOuts(repDimNames) *
                           warpsDotOperand(ctx, warpShape, warpOrder, kDimIdx)
                               .transposeOuts(repDimNames);
```

in second variant layout is still extends beyond K dimension due to lane component. to make it uniform I can introduce something like `laneDotOperand`, but this function will be used only in one place.

Reply:
I see it now. But if you use `warpsDotOperand`, do you still need expanding the thread size?

Maybe the following lines can be taken out

```
  auto kDimIdx = operandLayout.getOpIdx() == 0 ? rank - 1 : rank - 2;
  threadSize[kDimIdx] = shape[kDimIdx];
```

Reply:
> do you still need expanding the thread size?

No, technically, there are no need for this now. But this will generate different layout, in which it is impossible to layout values across k dimension. 

I would like to keep current implementation with manually extended K dimension, because with it we can use different order of elements in "per thread" tile.

Consider following example with more than one element in "non-k" dimension of "per thread size":

parent = blocked(sizePerThread=[2, 2], threads = [2, 2], warps = [1, 1], order = [1, 0], repeat order= [1,0])
dotOperand = dot(idx=0, parent)

Here number of threads = 4 for simplicity.

Let's say we have a tensor [4, 4]

### Current implementation

It will position register in following fashion:

#### per- thread layout
![Screenshot from 2024-12-20 22-19-25](https://github.com/user-attachments/assets/88103bef-140b-4e94-a242-1caa557b35df)

Here we are free to choose any order of elements we want, it could be k-first/mn-first/batch-first.

#### per-warp/cta layout and cta layout(cta is same, since in example num_warps=1)

![Screenshot from 2024-12-20 22-04-48](https://github.com/user-attachments/assets/1db96dcd-6e3d-4e22-b0d2-e2869671f7bf)

### Leave K expansion to `combineCtaCgaWithShape`

#### per-thread layout
![Screenshot from 2024-12-20 22-22-19](https://github.com/user-attachments/assets/1c78d0c3-5513-4c56-8f61-f54c858c0555)

### per-warp/cta layout
![Screenshot from 2024-12-20 22-07-36](https://github.com/user-attachments/assets/51b66064-f74c-4ca7-8f2d-ed6882f3819f)

#### extend it with repeats
![Screenshot from 2024-12-20 22-07-28](https://github.com/user-attachments/assets/7aaa64e1-8b88-4740-b8e5-e93ac5d29247)

At this point algorithm already finished working with cta layout and there are only one way to expand it across K, which places register in unexpected order. We can not make it k-first.

Reply:
Note that sizePerThread=[2, 2].

How would this picture only contain two elements?

![image](https://github.com/user-attachments/assets/1ff2df3d-5677-4d3b-bdd6-ad9e26295be5)

Reply:
I think the per thread layout would be

[r0, r1]
[r2, r3]

Reply:
Oh, [2*2] is a blocked parent layout property.

Dot operand is slightly different. It inherits all attributes of a parent, except k dimension. Dot operand layout implicitly extends per thread size to [2, K] for A operand and [K, 2] for B operand

Reply:
Picture you mentioned is related to intermediate layout, before it is expanded with com ineCtaCgaWithShape

Reply:
Got it. I think it's different from the cases where the parent is mma; we don't do implicit casting on only the register dimension. That being said, I think some code clean ups have to happen later. Right now, several methods crash on this dotoperand<blockedLayout>, like `getElemsPerThread`

Reply:
The update looks good to me now. Thanks!